client/src/entrypoints/admin/bulk-actions.js

```diff
@@ -28,59 +28,66 @@ function SelectBulkActionsFilter(e) {
 /* Event listener for individual page checkbox */
 function SelectBulkActionsCheckboxes(e) {
   if (checkedState.selectAllInListing) checkedState.selectAllInListing = false;
-  const prevLength = checkedState.checkedPages.size;
-  if (e.target.checked) checkedState.checkedPages.add(+e.target.dataset.pageId);
+  const prevLength = checkedState.checkedObjects.size;
+  if (e.target.checked) checkedState.checkedObjects.add(+e.target.dataset.objectId);
   else {
     /* unchecks `Select all` checkbox as soon as one page is unchecked */
-    document.querySelector(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH} input`).checked = false;
-    checkedState.checkedPages.delete(+e.target.dataset.pageId);
+    document.querySelectorAll(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH} input`).forEach(_el => {
+      const el = _el;
+      el.checked = false;
+    });
+    checkedState.checkedObjects.delete(+e.target.dataset.objectId);
   }
 
-  if (checkedState.checkedPages.size === 0) {
+  if (checkedState.checkedObjects.size === 0) {
     /* when all checkboxes are unchecked */
-    document.querySelectorAll(`.${TABLE_HEADERS_TR} > th`).forEach(el => el.classList.remove('u-hidden'));
-    document.querySelector(`.${BULK_ACTION_CHOICES_DIV}`).classList.add('u-hidden');
+    document.querySelector(`.${BULK_ACTION_CHOICES_DIV}`).classList.add('hidden');
     document.querySelectorAll(`.${BULK_ACTION_PAGE_CHECKBOX_INPUT}`).forEach(el => el.classList.remove('show'));
-    document.querySelector(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH}`).setAttribute('colspan', '1');
-  } else if (checkedState.checkedPages.size === 1 && prevLength === 0) {
+  } else if (checkedState.checkedObjects.size === 1 && prevLength === 0) {
     /* when 1 checkbox is checked for the first time */
     document.querySelectorAll(`.${BULK_ACTION_PAGE_CHECKBOX_INPUT}`).forEach(el => {
       el.classList.add('show');
     });
-    document.querySelectorAll(`.${TABLE_HEADERS_TR} > th`).forEach(el => el.classList.add('u-hidden'));
-    document.querySelector(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH}`).classList.remove('u-hidden');
-    document.querySelector(`.${BULK_ACTION_CHOICES_DIV}`).classList.remove('u-hidden');
-    document.querySelector(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH}`).setAttribute('colspan', '6');
+    document.querySelector(`.${BULK_ACTION_CHOICES_DIV}`).classList.remove('hidden');
   }
 
-  if (checkedState.checkedPages.size === checkedState.numPages) {
+  if (checkedState.checkedObjects.size === checkedState.numObjects) {
     /* when all checkboxes in the page are checked */
-    document.querySelector(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH} input`).checked = true;
+    document.querySelectorAll(`.${BULK_ACTION_SELECT_ALL_CHECKBOX_TH} input`).forEach(_el => {
+      const el = _el;
+      el.checked = true;
+    });
+    if (checkedState.shouldShowAllInListingText) {
```

Comment:
discussed during catch up: when the number of items in the listing fits on one page, i.e. no pagination, the "Select all in listing" option should not appear as it is confusing.
The listing uses the paginator so you could check for next/prev page existence in https://github.com/wagtail/wagtail/pull/7340/files#diff-5ae4cc8a762e77181ca3abcd029bdd0e759928f484d7a72457749ad1b9980412R23 `pages.has_previous and pages.has_next`